src/aof.c

```diff
@@ -40,191 +40,603 @@
+ * The following code implements the read/write logic of AOF manifest file, which 
+ * is used to track and manage all AOF files.
  *
- * We only need to append, but can't just use realloc with a large block
- * because 'huge' reallocs are not always handled as one could expect
- * (via remapping of pages at OS level) but may involve copying data.
- *
- * For this reason we use a list of blocks, every block is
- * AOF_RW_BUF_BLOCK_SIZE bytes.
+ * There are three types of AOF, they are: 
+ * BASE: Every time AOFRW success, a BASE type aof will be generated, which 
+ *       represents the redis SNAPSHOT at the moment when the AOFRW is executed.
+ *       In the aof manifest file, the BASE type AOF (if we have) is always at the 
+ *       beginning of the file. there is at most one BASE AOF.
+ * HISTORY: Each time the AOFRW is successful, the previous BASE AOF and INCR 
+ *       AOFs will become HISTORY. they will be cleaned regularly in cron. 
+ * INCR: There may be more than one (after multiple AOFRW failures), and they 
+ *       together represent all the incremental commands executed by redis after
+ *       the last AOFRW.
+ * 
+ * The following is a possible aof manifest file content:
+ * 
+ * fileName appendonly.2.rdb fileSeq 2 fileType b
+ * fileName appendonly.1.aof fileSeq 1 fileType h
+ * fileName appendonly.2.aof fileSeq 2 fileType h
+ * fileName appendonly.3.aof fileSeq 3 fileType h
+ * fileName appendonly.4.aof fileSeq 4 fileType i
+ * fileName appendonly.5.aof fileSeq 5 fileType i
  * ------------------------------------------------------------------------- */
 
-#define AOF_RW_BUF_BLOCK_SIZE (1024*1024*10)    /* 10 MB per block */
-
-typedef struct aofrwblock {
-    unsigned long used, free, pos;
-    /* Note that 'buf' must be the last field of aofrwblock struct, because
-     * memory allocator may give us more memory than our apply for reducing
-     * fragments, but we want to make full use of given memory, i.e. we may
-     * access the memory after 'buf'. To avoid make others fields corrupt,
-     * 'buf' must be the last one. */
-    char buf[AOF_RW_BUF_BLOCK_SIZE];
-} aofrwblock;
-
-/* This function free the old AOF rewrite buffer if needed, and initialize
- * a fresh new one. It tests for server.aof_rewrite_buf_blocks equal to NULL
- * so can be used for the first initialization as well. */
-void aofRewriteBufferReset(void) {
-    if (server.aof_rewrite_buf_blocks)
-        listRelease(server.aof_rewrite_buf_blocks);
-
-    server.aof_rewrite_buf_blocks = listCreate();
-    listSetFreeMethod(server.aof_rewrite_buf_blocks,zfree);
+/* Naming rules */
+#define BASE_AOF_SUFFIX          ".rdb"        
+#define INCR_AOF_SUFFIX          ".aof"        
+#define MANIFEST_NAME_SUFFIX     ".manifest"   
+#define MANIFEST_TEM_NAME_PREFIX "temp_"      
+
+/* AOF manifest key */
+#define AOF_MANIFEST_KEY_FILE_NAME   "fileName"
+#define AOF_MANIFEST_KEY_FILE_SEQ    "fileSeq"
+#define AOF_MANIFEST_KEY_FILE_TYPE   "fileType"
```

Comment:
i think this the "file" prefix and camel-case may be redundant
```suggestion
#define AOF_MANIFEST_KEY_FILE_NAME   "file"
#define AOF_MANIFEST_KEY_FILE_SEQ    "seq"
#define AOF_MANIFEST_KEY_FILE_TYPE   "type"
```